### PR TITLE
Lumen: Test and fix external export routes

### PR DESCRIPTION
**Pull request by rjmackay**
This pull request makes the following changes:
- Add tests for all exports/jobs/external routes
- Fix all the bugs I found (lots)
- [x] Add tests for the signature auth layer

Test checklist:
- [ ] composer test

Ping @ushahidi/platform

### CODE REVIEW

File: src/App/Repository/Form/AttributeRepository.php [hunk 1]
```diff
@@ -245,20 +246,23 @@ public function getByForm($form_id)
...
-			INNER JOIN forms ON form_stages.form_id = forms.id
-			where forms.id IN :forms
-			ORDER BY form_stages.priority, form_attributes.priority
-		";
-		$results = DB::query(Database::SELECT, $sql)
-			->bind(':forms', $form_ids)
-			->execute($this->db);
-		$attributes = $results->as_array();
+        $attributes = [];
+        if (count($form_ids) > 0) { // @FIXME: how would empty form_id even happen?
```
Comment (rowasc) [inline src/App/Repository/Form/AttributeRepository.php]: ohh. I guess this could happen if the is no data?

File: src/Console/Command/PostExporter.php [hunk 2]
```diff
@@ -68,15 +67,17 @@ public function handle()
...
 
         // Construct a Search Data object to hold the search info
         $data = $this->data->get('search');
 
         // Get CLI params
 		$limit = $this->option('limit');
         $offset = $this->option('offset');
-        $job_id = $input->getOption('job');
-        $add_header = $input->getOption('include_header');
+        $job_id = $this->option('job');
```
Comment (rowasc) [inline src/Console/Command/PostExporter.php]: I think this has spaces instead of tabs

File: src/Console/Command/PostExporter.php [hunk 3]
```diff
@@ -32,53 +32,54 @@
...
+	 * @var string
+	 */
+	protected $name = 'export';
+
+	/**
+	 * The console command signature.
+	 *
+	 * @var string
+	 */
+	protected $signature = 'export {--limit=100} {--offset=0} {--job} {--include-header=1}';
```
Comment (kinstelli) [inline src/Console/Command/PostExporter.php]: Are these meant to be defaults?

File: src/Init.php [hunk 4]
```diff
@@ -386,7 +386,9 @@ function feature($name)
 
 // Tools
 $di->set('tool.signer', $di->lazyNew('Ushahidi\Core\Tool\Signer'));
-$di->set('tool.verifier', $di->lazyNew('Ushahidi\Core\Tool\Verifier'));
+$di->set('tool.verifier', $di->lazyNew('Ushahidi\Core\Tool\Verifier', [
+	'apiKeyRepo' => $di->lazyGet('repository.apikey')
```
Comment (kinstelli) [inline src/Init.php]: Nice.

File: tests/unit/Core/Tool/VerifierTest.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
+<?php
+
+/**
+ * Unit tests for Lumen implementation of Ushahidi\Core\Tool\Mailer
```
Comment (kinstelli) [inline tests/unit/Core/Tool/VerifierTest.php]: `Verifier`, I think :)

File: tests/unit/App/Http/Middleware/SignatureAuthTest.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
+<?php
+
+/**
+ * Unit tests for Lumen implementation of Ushahidi\Core\Tool\Mailer
```
Comment (kinstelli) [inline tests/unit/App/Http/Middleware/SignatureAuthTest.php]: Also maybe outdated comment

File: tests/unit/Core/Tool/VerifierTest.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->checkApiKey($this->apiKey);
+
+		$this->assertTrue($return);
+	}
+}
```
Comment (kinstelli) [inline tests/unit/Core/Tool/VerifierTest.php]: It might be helpful to add test case(s) where the payload is an empty array, such as in a GET with long querystring, which is apparently also a valid situation, but caused me some grief once.

File: src/Console/Command/PostExporter.php (hunk 3 above)
Comment (rjmackay) [reply src/Console/Command/PostExporter.php]: Yup

Comment (rjmackay) [reply src/Console/Command/PostExporter.php]: Lumen command definitions are much easier

File: tests/unit/Core/Tool/VerifierTest.php (hunk 5 above)
Comment (rjmackay) [reply tests/unit/Core/Tool/VerifierTest.php]: oops

File: src/Core/Tool/Signer.php [hunk 8]
```diff
@@ -27,42 +27,19 @@ public function sign($fullUrl, $json)
...
         // Or http://pear.php.net/package/Crypt_HMAC/
         return base64_encode(hash_hmac("sha256", $url, $this->authToken, true));
     }
 
-    public function validate($expectedSignature, $url, $data = array())
+    public function validate($expectedSignature, $url, $data = "")
     {
-        return self::compare(
-            $this->computeSignature($url, json_encode($data)),
+        return hash_equals(
```
Comment (rjmackay) [inline src/Core/Tool/Signer.php]: @willdoran I'd really like you to check this. I've dropped the DIY comparison function in favour of PHP's built int.
The docs mention:

> Both arguments must be of the same length to be compared successfully. When arguments of differing length are supplied, FALSE is returned immediately and the length of the known string may be leaked in case of a timing attack.

However from my reading of `compare()` below, it looks like it exited out early if strlen was different so would has a similar timing vulnerability.

File: tests/unit/Core/Tool/VerifierTest.php (hunk 7 above)
Comment (rjmackay) [reply tests/unit/Core/Tool/VerifierTest.php]: Ok. Can do

File: tests/unit/Core/Tool/VerifierTest.php [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+		$signature = $this->makeSig($this->sharedSecret, $url, json_encode($payload));
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->verified($signature, $this->apiKey, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertTrue($return);
+	}
+
+	public function testValidSignatureWithEmptyPayload()
```
Comment (rjmackay) [inline tests/unit/Core/Tool/VerifierTest.php]: @kinstelli does this work?

File: tests/unit/Core/Tool/VerifierTest.php [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = "";
+		$signature = $this->makeSig($this->sharedSecret, $url, "");
```
Comment (kinstelli) [inline tests/unit/Core/Tool/VerifierTest.php]: My brain is dead now, but I think this actually fails if it's not a json_encoded empty array. 🤷‍♂️

Comment (rjmackay) [reply tests/unit/Core/Tool/VerifierTest.php]: So I changed the Signer to stop json_encoding everything... because thats weird. We now validate the raw contents of the request body as a string

File: src/Core/Tool/Signer.php [hunk 11]
```diff
@@ -27,42 +27,19 @@ public function sign($fullUrl, $json)
...
         // Note: hash_hmac requires PHP 5 >= 5.1.2 or PECL hash:1.1-1.5
         // Or http://pear.php.net/package/Crypt_HMAC/
         return base64_encode(hash_hmac("sha256", $url, $this->authToken, true));
     }
 
-    public function validate($expectedSignature, $url, $data = array())
+    public function validate($expectedSignature, $url, $data = "")
     {
-        return self::compare(
-            $this->computeSignature($url, json_encode($data)),
```
Comment (rjmackay) [inline src/Core/Tool/Signer.php]: @kinstelli note I dropped the `json_encode` here... so we expect request body as a string

File: app/Http/Middleware/SignatureAuth.php [hunk 12]
```diff
@@ -22,9 +30,13 @@ public function handle($request, Closure $next)
...
 
-        $verifier = new Verifier($signature, $api_key, $shared_secret, $fullURL, $request->getInputSource()->all());
-
-        if (!$verifier->verified()) {
+        if (!$this->verifier->verified(
+            $signature,
+            $api_key,
+            $shared_secret,
+            $fullURL,
+            $request->getContent()
```
Comment (rjmackay) [inline app/Http/Middleware/SignatureAuth.php]: This is now the string request contents, not an array decoded from json

Comment (kinstelli) [reply app/Http/Middleware/SignatureAuth.php]: Oh, Missed that. That's good.

File: tests/unit/Core/Tool/VerifierTest.php (hunk 7 above)
Comment (rowasc) [reply tests/unit/Core/Tool/VerifierTest.php]: tiny comment in case it helps: $data was by default an array (when doing the compute signature) so that's why we got the weird empty array appended to our urls even in empty get requests.

File: src/Core/Tool/Signer.php (hunk 8 above)
Comment (willdoran) [reply src/Core/Tool/Signer.php]: I think you're correct, the one below does have the same issue, I'm not sure why the note says that it doesn't. We can either go with the hash_equals and note the issue or we can implement one of the time insensitive versions.
